### PR TITLE
Installation and Adoption Fixes

### PR DESCRIPTION
**Pull request by camagallen**
# Pull Request Submission

## About
The goal of these updates is to smooth the installation and startup experience for the tool

Resolves: [AB#12907](https://dev.azure.com/kensci/91234890-e1c9-4679-971f-09afd8516db1/_workitems/edit/12907), [AB#12998](https://dev.azure.com/kensci/91234890-e1c9-4679-971f-09afd8516db1/_workitems/edit/12998), [AB#13008](https://dev.azure.com/kensci/91234890-e1c9-4679-971f-09afd8516db1/_workitems/edit/13008)

## Integration Issues
Due to our continued reliance on AIF360 for many measures, some installation issues could not be resolved. However, users have been directed toward the AIF360 troubleshooting documentation to resolve those issues.

Also, due to continued reliance on MIMIC data, some notebooks were made to escape integrated testing in order to reduce activation requirements for development.

## Checklist
- [X] I have read the [Contributing](CONTRIBUTING.md) guidelines.
- [X] I performed a self-review of my own code.
- [X] The CHANGELOG and version.json have been updated.
- [X] Documentation has been updated to reflect the changes.
    - [X] The code has clear and helpful comments, particularly in hard-to-understand areas.
    - [X] READMEs, Tutorials/Examples, and links have been updated as appropriate.
- [X] Tests have been added to prove that the fix is effective or that the feature works.
- [X] Pytest was run; new and existing unit tests pass locally with my changes.
- [X] I have made note of any breakages or dependencies (if applicable).
- [X] Any dependent changes have been merged and published in downstream modules.
- [X] These changes generate no new warnings.

## How was this update tested?
Tool installation was tested on a blank environment in both Windows 10 and macOS Big Sur. Pytest was also successfully run to completion on clones of the branch for both operating systems. 

## Other Relevant Information
@kenviku 
The objective of this update is to reduce barriers to installation and adoption. The largest change is the removal of Fairlearn as a back-end requirement since it frequently led to installation issues internally and forced a narrow band of version compatibility for the library's dependencies (i.e., the number of possible combinations of dependency versions was very small). Removing Fairlearn meant adding/updating two private modules (__classification_metrics and __fairness_metrics), which had touchpoints throughout the library. The process of integrating these two new modules also revealed a number of minor bugs, which were addressed in the spirit of reducing barriers to adoption.

### Files affected by removing Fairlearn
 * fairmlhealth/__classification_metrics.py (at line 32; other changes are refactoring in support of the update)
 * fairmlhealth/__fairness_metrics.py
 * fairmlhealth/reports.py
 * fairmlhealth/stratified_reports.py
 * fairmlhealth/tutorial_helpers.py (includes minor code cleanup)
 * tutorials_and_examples/*. (added validation requirement and re-ran tutorial notebooks)
 * test/test__classificationmetrics.py
 * test/test__fairnessmetrics.py
 * setup.py
 * utils.py
 * README.md

### Files affected by updates that facilitate adoption by facilitating external development
* test/notebook_tester.py
* test/test_tutorial_evaluatingbinaryclassification.py
* fairmlhealth/mimic_data.py (updated *.isna() calls for better compatibility with newer versions of pandas)
* docs/development/PULL_REQUEST_TEMPLATE.md
* docs/development/ISSUE_TEMPLATE.md

### Files affected by bug corrections (issues either errantly introduced on merge with recent integration branch updates or discovered thanks to improved testing)
 * README.md
 * tutorials_and_examples/*. (typos corrected and notebooks re-run)
 * fairmlhealth/stratified_reports.py
 * test/test_modelcomparison.py
 * test/development/model_comparison_debug.py 
 * fairmlhealth/model_comparison.py 
 * docs/publications/KDD2020-FairnessInHealthcareML-TutorialNotebook.ipynb
 * docs/Evaluating_Fairness.md

### CODE REVIEW

File: fairmlhealth/utils.py [hunk 1]
```diff
@@ -148,12 +150,26 @@ def __validate_report_input(X, y_true=None, y_pred=None, y_prob=None,
...
+            raise ValueError(err)
     # priv_grp
     if not isinstance(priv_grp, int):
         raise TypeError("priv_grp must be an integer")
 
     # If all above runs, inputs pass validation
     return True
+
+
+def validate_notebook_requirements():
```
Comment (camagallen) [inline fairmlhealth/utils.py:164]: create issue to update this feature to prompt user to install tutorial requirements, then remove tutorial requirements from default install.

File: Dockerfile [hunk 2]
```diff
@@ -9,7 +9,6 @@ COPY . .
 RUN python3 -m pip install --upgrade wheel setuptools pip
 RUN python3 setup.py install
 RUN python3 -m pip install -U pytest
-RUN python3 -m pip install -U nbformat nbconvert ipython
```
Comment (camagallen) [inline Dockerfile:12]: line introduced in error

**Review by kenviku — COMMENTED** [review]
Some nit comments. Mostly looks good!

File: fairmlhealth/stratified_reports.py [hunk 3]
```diff
@@ -90,11 +96,11 @@ def __preprocess_stratified(X, y_true, y_pred=None, y_prob=None,
     #
     over_max_vals = []
     for f in stratified_features:
-        # stratified analysis can only be run on discrete columns
         if df[f].nunique() > max_cats:
             over_max_vals.append(f)
-        elif df[f].isna().any():
-            df[f].fillna(np.nan, inplace=True)
+        else:
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairmlhealth/stratified_reports.py:101]: removed in next PR

File: fairmlhealth/stratified_reports.py [hunk 4]
```diff
@@ -240,12 +247,14 @@ def classification_performance(X, y_true, y_pred, y_prob=None,
...
-            # raise ValueError(f"Error processing {f}. {e}\n")
-            print(f"Error processing {f}. {e}\n")
+            errs[f] = e
             continue
         r = r.reset_index().rename(columns={f: 'FEATURE VALUE'})
         r.insert(0, 'FEATURE', f)
         res.append(r)
+    if any(errs):
+        for k, v in errs.items():
+            print(f"Error processing column(s) {k}. {v}\n")
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairmlhealth/stratified_reports.py:257]: will correct in next PR

File: fairmlhealth/stratified_reports.py [hunk 5]
```diff
@@ -522,13 +516,15 @@ def regression_fairness(X, y_true, y_pred, features:list=None, **kwargs):
...
-            print(f"Error processing {f}. {e}\n")
+            errs[f] = e
             continue
         r = pd.DataFrame(meas, index=[0])
         r['FEATURE'] = f
         r['FEATURE VALUE'] = v
         res_f.append(r)
+    if any(errs):
+        for k, v in errs.items():
+            print(f"Error processing column(s) {k}. {v}\n")
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairmlhealth/stratified_reports.py:527]: resolved in next PR

File: fairmlhealth/__classification_metrics.py [hunk 6]
```diff
@@ -1,22 +1,39 @@
...
+__all__ = ["accuracy", "binary_prediction_results", "balanced_accuracy",
+           "false_negative_rate", "false_positive_rate", "f1_score",
+           "negative_predictive_value", "roc_auc_score", "precision",
+           "pr_auc_score", "precision", "true_negative_rate",
+           "true_positive_rate"]
+
+
+''' Utilities '''
 
 def binary_prediction_results(y_true, y_pred):
```
Comment (camagallen) [inline fairmlhealth/__classification_metrics.py:21]: We do use the classification_report elsewhere. For this we need TP, TN, FP, and FN, which we will not get from classification_report. However, this function **is** updated to use scikit's confusion_matrix in the next PR.

[the remark this answers is not in the document]
Comment (camagallen) [reply fairmlhealth/__classification_metrics.py:21]: We do use the classification_report elsewhere. For this we need TP, TN, FP, and FN, which we will not get from classification_report. However, this function **is** updated to use scikit's confusion_matrix in the next PR.

**Review by kenviku — APPROVED** [review]
LGTM

**Pull request merged**